app/scripts/services/authorization.js

```diff
@@ -0,0 +1,105 @@
+            normalizedRules[(apiGroup === "") ? resource : apiGroup + "/" + resource] = rule.verbs;
+          });
+
+          if (!canAddToProject && checkResource(resource) && (!_.isEmpty(_.intersection(rule.verbs ,(["*", "create", "update"]))))) {
+            canAddToProject = true;
+          }
+        });
+      });
+      return [normalizedRules, canAddToProject];
+    };
+
+    AuthorizationService.prototype.getProjectRules = function(projectName) {
+      var self = this;
+      var deferred = $q.defer();
+      this.currentProject = projectName;
+      var projectRules = self.cachedRulesByProject.get(projectName);
+      if (!projectRules || projectRules.forceRefresh) {
+        Logger.log("AuthorizationService, loading user rules for " + projectName + " project");
+        var object = {kind: "SelfSubjectRulesReview",
+                      apiVersion: "v1"
+                    };
+        DataService.create('selfsubjectrulesreviews', null, object, {namespace: projectName}).then(
+          function(data) {
+            var normalizedData = normalizeRules(data.status.rules);
+            self.cachedRulesByProject.put(projectName, {rules: normalizedData[0],
+                                                        canAddToProject: normalizedData[1],
+                                                        forceRefresh: false,
+                                                        cacheTimestamp: _.now()
+                                                      });
+            deferred.resolve(true);
+          }, function() {
+            deferred.reject(false);
+        });
+      } else {
+        // Using cached data.
+        Logger.log("AuthorizationService, using cached rules for " + projectName + " project");
+        if ((_.now() - projectRules.cacheTimestamp) >= 600000) {
+          projectRules.forceRefresh = true; 
+        }
+        deferred.resolve(true);
+      }
+      return deferred.promise;
+    };
+
+    AuthorizationService.prototype.canI = function(verb, resource, projectName) {
+      projectName = projectName || this.currentProject;
+      var rules = this.getRulesForProject(projectName);
+      if (rules) {
+        if (rules[resource]) {
+          return _.contains(rules[resource], verb) || _.contains(rules[resource], '*');
+        } else if (rules['*']) {
+          return _.contains(rules['*'], verb) || _.contains(rules['*'], '*');
+        } else {
+          return false;
+        }    
+      } else {
+        return false;
+      }
+    };
+
+    AuthorizationService.prototype.canIAddToProject = function(projectName) {
+      return !!_.get(this.cachedRulesByProject.get(projectName || this.currentProject), ['canAddToProject']);
+    };
+
+
+    AuthorizationService.prototype.getRulesForProject = function(projectName) {
+      return _.get(this.cachedRulesByProject.get(projectName || this.currentProject), ['rules']);
+    };
+
+    return new AuthorizationService();
```

Comment:
I don't think anyone else cares too much, but I still think we should shy away from the trouble of setting up a `prototype` but then calling the constructor and returning a singleton anyway.  I think our services code can be cleaner if we just `return { /* hash of functions */ }`